int64 th
int64 x
int64 y
int64 thL
int64 thR
---
int64 th 
int64 x
int64 y
int64 thL
int64 thR